File: azkar/kotlin/shared/src/commonMain/kotlin/com/wildduck/azkar/app/PlanFile.kt
// plan.json: the reminders iOS has already been handed, so a later launch knows which of them have fired.
package com.wildduck.azkar.app

import com.wildduck.azkar.core.Card
import com.wildduck.azkar.core.PlannedReminder
import com.wildduck.azkar.core.Session
import com.wildduck.azkar.core.Zikr
import kotlin.time.Instant
import kotlinx.serialization.SerializationException
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
private data class StoredZikr(val text: String, val count: Int = 1, val note: String? = null)

@Serializable
private data class StoredReminder(
    /** The time in UTC, to the second: "2026-09-16T18:45:00Z". */
    val at: String,
    val session: String,
    val position: Int? = null,
    val total: Int? = null,
    val zikr: StoredZikr,
    val state: StoredState = StoredState(),
)

@Serializable
private data class StoredPlan(val reminders: List<StoredReminder> = emptyList())

object PlanFile {
    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        explicitNulls = false
        encodeDefaults = true
    }

    fun encode(plan: List<PlannedReminder>): String = json.encodeToString(
        StoredPlan.serializer(),
        StoredPlan(
            plan.map { reminder ->
                StoredReminder(
                    at = Instant.fromEpochSeconds(reminder.fireAt.epochSeconds).toString(),
                    session = reminder.card.session.name.lowercase(),
                    position = reminder.card.position,
                    total = reminder.card.total,
                    zikr = reminder.card.zikr.let { StoredZikr(it.text, it.count, it.note) },
                    state = reminder.state.stored(),
                )
            },
        ),
    )

    /** A plan is a convenience, not a record: anything unreadable is simply no plan. */
    fun decode(text: String?): List<PlannedReminder> {
        val stored = try {
            text?.let { json.decodeFromString(StoredPlan.serializer(), it) }
        } catch (_: SerializationException) {
            null
        } ?: return emptyList()
        return stored.reminders.mapNotNull { reminder ->
            val fireAt = try {
                Instant.parse(reminder.at)
            } catch (_: IllegalArgumentException) {
                return@mapNotNull null
            }
            val session = Session.entries.firstOrNull { it.name.lowercase() == reminder.session }
                ?: return@mapNotNull null
            PlannedReminder(
                fireAt = fireAt,
                card = Card(
                    zikr = reminder.zikr.let { Zikr(text = it.text, count = it.count, note = it.note) },
                    session = session,
                    position = reminder.position,
                    total = reminder.total,
                ),
                state = reminder.state.state(),
            )
        }
    }
}

File: azkar/apple/Core/PlanFile.swift
// plan.json: the reminders iOS has already been handed, so a later launch knows which of them have fired.
import Foundation

private struct StoredZikr: Codable {
  var text: String
  var count: Int
  var note: String?
}

private struct StoredReminder: Codable {
  var at: String
  var session: Session
  var position: Int?
  var total: Int?
  var zikr: StoredZikr
  var state: AppState
}

private struct StoredPlan: Codable {
  var reminders: [StoredReminder]
}

enum PlanFile {
  /// Whole seconds in UTC — "2026-09-16T18:45:00Z" — which is how Kotlin writes an instant.
  private static let iso: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime]
    f.timeZone = TimeZone(identifier: "UTC")
    return f
  }()

  private static let isoWithFraction: ISO8601DateFormatter = {
    let f = ISO8601DateFormatter()
    f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    f.timeZone = TimeZone(identifier: "UTC")
    return f
  }()

  static func encode(_ plan: [PlannedReminder]) -> String {
    let stored = StoredPlan(
      reminders: plan.map { reminder in
        StoredReminder(
          at: iso.string(from: Date(timeIntervalSince1970: reminder.fireAt.timeIntervalSince1970.rounded(.down))),
          session: reminder.card.session,
          position: reminder.card.position,
          total: reminder.card.total,
          zikr: StoredZikr(
            text: reminder.card.zikr.text, count: reminder.card.zikr.count, note: reminder.card.zikr.note),
          state: reminder.state)
      })
    guard let data = try? JSONEncoder().encode(stored), let text = String(data: data, encoding: .utf8) else {
      return #"{"reminders":[]}"#
    }
    return text
  }

  /// A plan is a convenience, not a record: anything unreadable is simply no plan.
  static func decode(_ text: String?) -> [PlannedReminder] {
    guard let data = text?.data(using: .utf8),
      let stored = try? JSONDecoder().decode(StoredPlan.self, from: data)
    else { return [] }
    return stored.reminders.compactMap { reminder in
      guard let fireAt = iso.date(from: reminder.at) ?? isoWithFraction.date(from: reminder.at) else { return nil }
      return PlannedReminder(
        fireAt: fireAt,
        card: Card(
          zikr: Zikr(text: reminder.zikr.text, count: reminder.zikr.count, note: reminder.zikr.note, ref: nil),
          session: reminder.session, position: reminder.position, total: reminder.total),
        state: reminder.state)
    }
  }
}
